elyra/pipeline/validate.py

```diff
@@ -0,0 +1,631 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+from enum import IntEnum
+from glob import glob
+import json
+import os
+import re
+from typing import Dict
+from typing import List
+from typing import Optional
+
+import networkx as nx
+from traitlets.config import SingletonConfigurable
+
+from elyra.pipeline.component import Component
+from elyra.pipeline.component_registry import ComponentRegistry
+from elyra.pipeline.parser import PipelineParser
+from elyra.pipeline.pipeline import Operation
+from elyra.pipeline.processor import PipelineProcessorManager
+
+
+class ValidationSeverity(IntEnum):
+    Error = 1
+    Warning = 2
+    Information = 3
+    Hint = 4
+
+
+class ValidationResponse(object):
+    def __init__(self):
+        self._response = {"title": "Elyra Pipeline Diagnostics",
+                          "description": "Issues discovered when parsing the pipeline",
+                          "issues": []
+                          }
+        self._has_fatal = False
+
+    @property
+    def response(self) -> Dict:
+        """
+        :return: The dict of validation errors and warnings found in the pipeline
+        """
+        return self._response
+
+    @property
+    def has_fatal(self):
+        return self._has_fatal
+
+    def add_message(self,
+                    message: str,
+                    message_type: Optional[str] = "",
+                    data: Optional[Dict] = "",
+                    severity: ValidationSeverity = ValidationSeverity.Warning):
+        """
+        Helper function to add a diagnostic message to the response to be sent back
+        :param message: A simple message describing the issue
+        :param message_type: The type of message to send back e.g. invalidNodeType, invalidPipeline
+        :param data: a Dict with granular details regarding the error e.g. the nodeID, pipelineID, linkID etc.
+        :param severity: the severity level of the issue
+        :return:
+        """
+        valid_severity_levels = [ValidationSeverity.Error, ValidationSeverity.Warning,
+                                 ValidationSeverity.Information, ValidationSeverity.Hint]
+
+        if severity in valid_severity_levels:
+            diagnostic = {"severity": severity.value,
+                          "source": "Elyra Pipeline Validation Service",
+                          "type": message_type,
+                          "message": message,
+                          "data": data
+                          }
+            self._response['issues'].append(diagnostic)
+
+        if severity is ValidationSeverity.Error:
+            self._has_fatal = True
+
+    def to_json(self):
+        return self._response
+
+
+class PipelineValidationManager(SingletonConfigurable):
+
+    def __init__(self, **kwargs):
+        super().__init__()
```

Comment:
We should remove this or propagate `**kwargs` to the superclass constructor.  I think the former is probably best since it's not doing anything anyway.

Reply:
just kept it and pushed to the superclass.